Firestore/core/src/firebase/firestore/util/executor_std.cc

```diff
@@ -0,0 +1,151 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/util/executor_std.h"
+
+#include <future>  // NOLINT(build/c++11)
+#include <sstream>
+
+namespace firebase {
+namespace firestore {
+namespace util {
+namespace internal {
+
+namespace {
+
+// The only guarantee is that different `thread_id`s will produce different
+// values.
+std::string ThreadIdToString(const std::thread::id thread_id) {
+  std::ostringstream stream;
+  stream << thread_id;
+  return stream.str();
+}
+
+}  // namespace
+
+ExecutorStd::ExecutorStd() {
+  // Somewhat counter-intuitively, constructor of `std::atomic` assigns the
+  // value non-atomically, so the atomic initialization must be provided here,
+  // before the worker thread is started.
+  // See [this thread](https://stackoverflow.com/questions/25609858) for context
+  // on the constructor.
+  current_id_ = 0;
+  shutting_down_ = false;
+  worker_thread_ = std::thread{&ExecutorStd::PollingThread, this};
+}
+
+ExecutorStd::~ExecutorStd() {
+  shutting_down_ = true;
+  // Make sure the worker thread is not blocked, so that the call to `join`
+  // doesn't hang.
+  UnblockQueue();
+  worker_thread_.join();
+}
+
+void ExecutorStd::Execute(Operation&& operation) {
+  DoExecute(std::move(operation), Immediate());
+}
+
+DelayedOperation ExecutorStd::Schedule(const Milliseconds delay,
+                                       TaggedOperation&& tagged) {
+  // While negative delay can be interpreted as a request for immediate
+  // execution, supporting it would provide a hacky way to modify FIFO ordering
+  // of immediate operations.
+  FIREBASE_ASSERT_MESSAGE(delay.count() >= 0,
+                          "Schedule: delay cannot be negative");
+
+  namespace chr = std::chrono;
+  const auto now = chr::time_point_cast<Milliseconds>(chr::system_clock::now());
+  const auto id =
+      DoExecute(std::move(tagged.operation), now + delay, tagged.tag);
+
+  return DelayedOperation{[this, id] { TryCancel(id); }};
+}
+
+void ExecutorStd::TryCancel(const Id operation_id) {
+  schedule_.RemoveIf(
+      [operation_id](const Entry& e) { return e.id == operation_id; });
+}
+
+ExecutorStd::Id ExecutorStd::DoExecute(Operation&& operation,
+                                       const TimePoint when,
+                                       const Tag tag) {
+  // Note: operations scheduled for immediate execution don't actually need an
+  // id. This could be tweaked to reuse the same id for all such operations.
+  const auto id = NextId();
+  schedule_.Push(Entry{std::move(operation), id, tag}, when);
+  return id;
+}
+
+void ExecutorStd::PollingThread() {
+  while (!shutting_down_) {
+    Entry entry = schedule_.PopBlocking();
+    if (entry.tagged.operation) {
+      entry.tagged.operation();
```

Comment:
I'm not sure if it's necessary to wrap this call in a mutex (`dispatch_async`/etc. operate as if under a locked mutex).